### PR TITLE
Support complex types in Arrow/Parquet/ORC

### LINKED ISSUE #17240: The type "dictionary" ... is not supported for conversion from a Arrow data format
I would like to request support of dictionary type when reading from Arrow format.

I am loading table using data from Arrow as follows
```sh
clickhouse-client --query "CREATE TABLE $2 (id1 String, id2 String, id3 String, id4 Int32, id5 Int32, id6 Int32, v1 Int32, v2 Int32, v3 Float64) ENGINE = MergeTree() ORDER BY tuple();"
clickhouse-client --max_memory_usage $CH_MEM --max_insert_threads 1 --query "INSERT INTO $2 SELECT id1, id2, id3, id4, id5, id6, v1, v2, v3 FROM file('data/$2.feather', 'Arrow', 'id1 String, id2 String, id3 String, id4 Int32, id5 Int32, id6 Int32, v1 Int32, v2 Int32, v3 Float64')"
```
and this results into
```
Received exception from server (version 20.9.3):
Code: 70. DB::Exception: Received from localhost:9000. DB::Exception: The type "dictionary" of an input column "id1" is not supported for conversion from a Arrow data format: While executing File. 
```
My CH version 20.9.3.

Note that when I change format to `CSVWithNames` and file extension to `csv` to read the same dataset but from the different format, it does work fine.

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support structs and maps in Arrow/Parquet/ORC and dictionaries in Arrow input/output formats. Present new setting `output_format_arrow_low_cardinality_as_dictionary`.

Detailed description / Documentation draft:
Support Struct and Map types in input/output column-oriented formats Arrow/Parquet/ORC. Now you can input/output ClickHouse Tuples and Maps (experimental type) in these formats. Nested complex types are also supported. If setting `output_format_arrow_low_cardinality_as_dictionary` is true, LowCardinality columns will be converted into dictionary Arrow column, if false, LowCardinality columns will be converted to full column before output. By default this setting is false.
Closes https://github.com/ClickHouse/ClickHouse/issues/17240 and https://github.com/ClickHouse/ClickHouse/issues/21866

### CODE REVIEW

**Comment by Avogar** [comment]
@Mergifyio update

File: src/Processors/Formats/Impl/ArrowBlockInputFormat.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #if USE_ARROW
 
 #include <Processors/Formats/IInputFormat.h>
+#include <Processors/Formats/Impl/ArrowColumnToCHColumn.h>
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ArrowBlockInputFormat.h:6]: I think importing Impl dir in .h file is not good, but ArrowColumnToCHColumn (and CHColumnToArrowColumn) became statefull and I need to create their instances as class members. Maybe move these classes into another dir (I am not sure which one)?

File: src/Core/Settings.h [hunk 2]
```diff
@@ -115,7 +115,7 @@ class IColumn;
     M(Bool, skip_unavailable_shards, false, "If 1, ClickHouse silently skips unavailable shards and nodes unresolvable through DNS. Shard is marked as unavailable when none of the replicas can be reached.", 0) \
     \
     M(UInt64, parallel_distributed_insert_select, 0, "Process distributed INSERT SELECT query in the same cluster on local tables on every shard, if 1 SELECT is executed on each shard, if 2 SELECT and INSERT is executed on each shard", 0) \
-    M(UInt64, distributed_group_by_no_merge, 0, "If 1, Do not merge aggregation states from different servers for distributed queries (shards will process query up to the Complete stage, initiator just proxies the data from the shards). If 2 the initiator will apply ORDER BY and LIMIT stages (it is not in case when shard process query up to the Complete stage)", 0) \
+    M(UInt64, distributed_group_by_no_merge, 0, "If 1, Do not merge aggregation states from different servers for distributed query processing - in case it is for certain that there are different keys on different shards. If 2 - same as 1 but also apply ORDER BY and LIMIT stages", 0) \
```
Comment (nikitamikhaylov) [inline src/Core/Settings.h:118]: I think these changes are not related to this PR.

Comment (Avogar) [reply src/Core/Settings.h:118]: It was a bad merge with master, I will fix it

**Comment by Avogar** [comment]
I am also going to add Map type support in this PR.

**Comment by Avogar** [comment]
@Mergifyio update

File: src/Processors/Formats/Impl/ArrowBlockInputFormat.h (hunk 1 above)
Comment (nikitamikhaylov) [reply src/Processors/Formats/Impl/ArrowBlockInputFormat.h:6]: Maybe better to use `pimpl` here? And move this include to the .cpp file.

Comment (Avogar) [reply src/Processors/Formats/Impl/ArrowBlockInputFormat.h:6]: That's a good idea, thanks

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 3]
```diff
@@ -260,19 +267,43 @@ namespace DB
...
+                throw Exception("Unsupported type for indexes in LowCardinality: " + arrow_column->type()->name() + ".", ErrorCodes::BAD_ARGUMENTS);
+        }
+    }
 
-    static void readColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> & arrow_column, IColumn & internal_column, const std::string & column_name, const std::string format_name, bool is_nullable)
+    static void readColumnFromArrowColumn(
+        std::shared_ptr<arrow::ChunkedArray> & arrow_column,
+        IColumn & internal_column,
+        const std::string & column_name,
+        const std::string format_name,
```
Comment (nikitamikhaylov) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:292]: Why not `const std::string &` ? It is confusing in comparison to the previous line.

Comment (Avogar) [reply src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:292]: It's a typo, will fix it

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 4]
```diff
@@ -304,10 +335,9 @@ namespace DB
                 //fillColumnWithNumericData<Decimal128, ColumnDecimal<Decimal128>>(arrow_column, read_column); // Have problems with trash values under NULL, but faster
```
Comment (nikitamikhaylov) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:335]: delete?

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 5]
```diff
@@ -369,6 +458,53 @@ namespace DB
...
+            if (!tuple_type)
+                throw Exception{"Cannot convert arrow STRUCT type to a not Tuple ClickHouse type " + column_type->getName(), ErrorCodes::CANNOT_CONVERT_TYPE};
+
+            const DataTypes & tuple_nested_types = tuple_type->getElements();
+            int internal_fields_num = tuple_nested_types.size();
+            /// If internal column has less elements then arrow struct, we will select only first internal_fields_num columns.
+            if (internal_fields_num > struct_type->num_fields())
+                throw Exception
+                    {
+                        "Cannot convert arrow STRUCT with " + std::to_string(struct_type->num_fields()) + " fields to a ClickHouse Tuple with "
```
Comment (nikitamikhaylov) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:474]: Consider to use `fmt::format`

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by Avogar** [comment]
@Mergifyio update

**Comment by Avogar** [comment]
@buyology thank you for you comment, I will try to support inserting into Nested by struct of arrays or arrays of struct in the next PR.

**Comment by Avogar** [comment]
AST fuzzer failure: https://github.com/ClickHouse/ClickHouse/issues/25293

**Comment by Avogar** [comment]
@nikitamikhaylov It seems I need some help, because I cannot understand special build failure. It seems fully unrelated to this PR but I don't see such failures anywhere else.

**Comment by Avogar** [comment]
@Mergifyio update

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-10557.

**Comment by alexey-milovidov** [comment]
Continued here: #36832
